File: python/kyrazo/webhook.py
import hmac
import hashlib
import time
from typing import Dict, Optional, Union
from .core.exceptions import ValidationError

class Webhook:
    """
    Webhook utilities for verifying authenticity of incoming requests.
    """
    
    HEADERS = {
        "SIGNATURE": "X-Webhook-Signature",
        "TIMESTAMP": "X-Webhook-Timestamp",
        "NONCE": "X-Webhook-Nonce",
        "VERSION": "X-Webhook-Version",
    }

    @staticmethod
    def verify(
        payload: str,
        headers: Dict[str, Union[str, list]],
        secret: str,
        tolerance: int = 300,
    ) -> bool:
        """
        Verify a webhook signature.
        
        Args:
            payload: The raw request body string.
            headers: The incoming request headers.
            secret: The signing secret for the target.
            tolerance: Maximum allowed age of the request in seconds (default: 300).
            
        Returns:
            True if the signature is valid.
            
        Raises:
            ValidationError: If verification fails.
        """
        # Normalize header keys to PascalCase or lowercase for robustness
        header_map = {k.lower(): v for k, v in headers.items()}
        
        sig_header = header_map.get(Webhook.HEADERS["SIGNATURE"].lower())
        timestamp = header_map.get(Webhook.HEADERS["TIMESTAMP"].lower())
        nonce = header_map.get(Webhook.HEADERS["NONCE"].lower())
        version = header_map.get(Webhook.HEADERS["VERSION"].lower())

        if not all([sig_header, timestamp, nonce, version]):
            raise ValidationError("Missing required webhook headers")

        if isinstance(sig_header, list): sig_header = sig_header[0]
        if isinstance(timestamp, list): timestamp = timestamp[0]
        if isinstance(nonce, list): nonce = nonce[0]
        if isinstance(version, list): version = version[0]

        # Verify timestamp tolerance to prevent replay attacks
        try:
            ts = int(timestamp)
        except ValueError:
            raise ValidationError("Invalid webhook timestamp format")
            
        now = int(time.time())
        if abs(now - ts) > tolerance:
            raise ValidationError("Webhook timestamp expired or invalid")

        # signature format is "v1={hash}"
        if "=" not in sig_header:
            raise ValidationError("Invalid webhook signature format")
            
        header_version, signature = sig_header.split("=", 1)
        if header_version != "v1" or not signature:
            raise ValidationError("Invalid webhook signature version or format")

        # Construct payload: version.timestamp.nonce.body
        signed_payload = f"{version}.{timestamp}.{nonce}.{payload}"

        # Compute expected HMAC
        expected_signature = hmac.new(
            secret.encode(),
            signed_payload.encode(),
            hashlib.sha256
        ).hexdigest()

        # Use timing-safe comparison
        if hmac.compare_digest(signature, expected_signature):
            return True

        raise ValidationError("Webhook signature verification failed")

File: javascript/src/webhook.ts
import * as crypto from "crypto";
import { ValidationError } from "./errors";

/**
 * Interface for webhook headers as received by a Node.js server (e.g. Express)
 * Note: Headers are typically lowercased by Node.js
 */
export interface WebhookHeaders {
  "x-webhook-signature"?: string;
  "x-webhook-timestamp"?: string;
  "x-webhook-nonce"?: string;
  "x-webhook-version"?: string;
  [key: string]: string | string[] | undefined;
}

/**
 * Webhook utilities for verifying authenticity of incoming requests
 */
export class Webhook {
  /**
   * Header names used by Kyrazo
   */
  public static readonly HEADERS = {
    SIGNATURE: "x-webhook-signature",
    TIMESTAMP: "x-webhook-timestamp",
    NONCE: "x-webhook-nonce",
    VERSION: "x-webhook-version",
  } as const;

  /**
   * Verify a webhook signature
   *
   * @param payload - The raw request body string
   * @param headers - The incoming request headers
   * @param secret - The signing secret for the target
   * @param tolerance - Maximum allowed age of the request in seconds (default: 300)
   * @returns true if the signature is valid
   * @throws ValidationError if verification fails
   */
  public static verify(
    payload: string,
    headers: WebhookHeaders,
    secret: string,
    tolerance: number = 300,
  ): boolean {
    const signatureHeader = headers[this.HEADERS.SIGNATURE];
    const timestamp = headers[this.HEADERS.TIMESTAMP];
    const nonce = headers[this.HEADERS.NONCE];
    const version = headers[this.HEADERS.VERSION];

    if (
      typeof signatureHeader !== "string" ||
      typeof timestamp !== "string" ||
      typeof nonce !== "string" ||
      typeof version !== "string"
    ) {
      throw new ValidationError("Missing required webhook headers");
    }

    // Verify timestamp tolerance to prevent replay attacks
    const now = Math.floor(Date.now() / 1000);
    const ts = parseInt(timestamp, 10);
    if (isNaN(ts) || Math.abs(now - ts) > tolerance) {
      throw new ValidationError("Webhook timestamp expired or invalid");
    }

    // signature format is "v1={hash}"
    const [headerVersion, signature] = signatureHeader.split("=");
    if (headerVersion !== "v1" || !signature) {
      throw new ValidationError("Invalid webhook signature format");
    }

    // Construct payload: version.timestamp.nonce.body
    const signedPayload = `${version}.${timestamp}.${nonce}.${payload}`;

    // Compute expected HMAC
    const expectedSignature = crypto
      .createHmac("sha256", secret)
      .update(signedPayload)
      .digest("hex");

    // Use timing-safe comparison
    try {
      if (
        crypto.timingSafeEqual(
          Buffer.from(signature),
          Buffer.from(expectedSignature),
        )
      ) {
        return true;
      }
    } catch (e) {
      // Buffer length mismatch or other error
    }

    throw new ValidationError("Webhook signature verification failed");
  }
}
